### PR TITLE
Hidden states extraction improvements

### PR DESCRIPTION
**Pull request by fynnsu**
## Purpose
The current ExampleHiddenStatesConnector saves the hidden states in `wait_for_save` method of the KVConnector. The challenge with this, is that when running with chunked prefill, these method is called multiple times per request. The current solution is to just disable chunked prefill and enforce that when using the connector.

This pr removes this requirement, by moving the save step into `get_finished`, which is only called once per request, after the response has been generated and the request has returned.

The new flow ordering is:
1. Request received by scheduler, `build_connector_meta` is called and populates `new_req_filenames`
2. In worker `save_kv_layer` (run after model forward). This is now a no-op
3. In worker `wait_for_save`, each new req that doesn't have a lock file gets one created and stored in a dict
4. In scheduler `request_finished`, creates a `PendingSave` object for the request. Returns hidden_states_path to client and delays block freeing.
5. In scheduler `build_connector_meta`, add the `PendingSave` objects to connector metadata to send to worker
6. In worker `get_finished`, read the pending saves and start async writes (releases lock when done)
8. In worker `get_finished` (later iteration), returns completed req ids so blocks can be freed and also cleans up lock files for failed/aborted requests
9. Client reads file after lock file is released.

I've also added a couple options to the connector that can be trigger by passing in `extra_args={"kv_transfer_params": ...}` to requests.

`hidden_states_path`: Allows callers to specify the path to save hidden states to (falls back to current behavior if not provided).
`include_output_tokens`: Allows callers to request hidden states/token ids for both the prompt+response (instead of just the prompt).

Both of these options can be specified per-request. I've updated `examples/features/speculative_decoding/extract_hidden_states_offline.py` to show usage and also updated docs. 

## Test Plan
Updated tests to no longer disable chunked prefill. Added a test for the new per-request options. 

## Test Result
Tests passed locally.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 1]
```diff
@@ -421,17 +406,25 @@ def build_connector_meta(
...
-            meta.add_request(
-                new_req.req_id,
-                filename=filename,
-                token_ids=token_ids,
+            kv_params = (
+                new_req.sampling_params.extra_args.get("kv_transfer_params")
+                if new_req.sampling_params and new_req.sampling_params.extra_args
+                else None
+            ) or {}
+            filename = kv_params.get(
```
[the remark this answers is not in the document]
Comment (fynnsu) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: Yeah, this make sense. Will fix.

[the remark this answers is not in the document]
Comment (fynnsu) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py]: Added an `allow_custom_save_path` arg when launching the server. If set False (default), ignore the custom paths and print a warning when one is received. If True, allow arbitrary save paths (also prints a warning on server launch).

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 2]
```diff
@@ -341,39 +309,56 @@ def save_kv_layer(
...
+        tensors = {
+            "hidden_states": pinned_hs,
+            "token_ids": pending.token_ids.clone(),
+        }
+
+        # Submit to thread pool for disk write.
+        prior = self._req_futures.get(pending.req_id)
+        assert prior is None, "Found another KV transfer request with same req_id!"
+
+        os.makedirs(os.path.dirname(pending.filename), exist_ok=True)
```
[the remark this answers is not in the document]
Comment (fynnsu) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:365]: Handled by the above.

**Comment by fynnsu** [comment]
@mgoin I'm pretty sure the failures are unrelated. (I checked ci.vllm.ai)

File: tests/v1/kv_connector/extract_hidden_states_integration/test_extraction.py
Comment (mgoin) [inline tests/v1/kv_connector/extract_hidden_states_integration/test_extraction.py:1]: @fynnsu Please add an explicit case to make sure we test that chunked prefill is actually run into. The trivial example would be to set max_num_batched_tokens to a small value like 128 and then send a few requests with 500 prompt tokens

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 3]
```diff
@@ -304,35 +285,31 @@ def save_kv_layer(
...
+
+        # Compute slot mapping from block_ids
+        block_ids_t = torch.tensor(pending.block_ids, dtype=torch.long)
+        num_blocks = block_ids_t.shape[0]
+        block_offsets = torch.arange(0, self._block_size, dtype=torch.long)
+        slot_mapping = (
+            block_offsets.reshape((1, self._block_size))
+            + block_ids_t.reshape((num_blocks, 1)) * self._block_size
         )
+        slot_mapping = slot_mapping.flatten()
```
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:321]: When is it safe to assume this contiguous ordering `block_offsets + block_ids * block_size`? It would be good if we had an assertion check that this is safe to assume. Not sure if hybrid models with mamba would disrupt this, or possibly cross-layer kv layout.

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 4]
```diff
@@ -444,35 +455,54 @@ def request_finished(
...
-        _ = self._active_requests.pop(req_id, None)
-        _ = self._req_blocks.pop(req_id, None)
-
-        return True, {"hidden_states_path": req_filename}
+        filename = self._request_filenames.pop(req_id)
+        kv_params = request.kv_transfer_params or {}
+        if kv_params.get("include_output_tokens", False):
+            # Exclude the final token — it was the model's output, never an
+            # input to a forward pass, so its hidden state is not in the cache.
+            token_ids = torch.tensor(list(request.all_token_ids)[:-1])
```
Comment (mgoin) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:478]: nit: with speculative_tokens>=1 and rejected drafts, are there edge cases where all_token_ids includes tokens that were drafted but rejected, so have no cached hidden state?

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 3 above)
Comment (fynnsu) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:321]: I assume you mean the ordering between blocks, because within a block the ordering should be just the automatic ordering.

Claude reply + my edits:

> The `block_id * block_size + offset` formula is the standard vLLM KV cache slot mapping — the same one used by all attention backends. Both the write and read sides use it, so they're consistent by construction:
> 
> **Write side**: During forward, the scheduler computes `slot_mapping` via `_compute_slot_mapping_kernel` in `block_table.py:378` (`slot_ids = block_numbers * block_size + local_block_offsets`), where `block_numbers` comes from the worker's block table. The block table is populated from `KVCacheBlocks.get_block_ids()` (`kv_cache_manager.py:84`). `CacheOnlyAttentionImpl.do_kv_cache_update` then writes hidden states into the KV cache at those slots via `basic_cache` (`extract_hidden_states.py:87`).
> 

This ^ is when we're inserting the hidden states into the kv cache.

> **Read side**: In `_connector_finished`, the scheduler calls `kv_cache_manager.get_block_ids(request_id)` — which calls the same `KVCacheBlocks.get_block_ids()` on the same `req_to_blocks` list. The connector reconstructs slot indices using the same formula.
> 

And then when extracting ^, we use the same `get_block_ids()` function. 

> So the block_ids on both sides come from the same `req_to_blocks: list[KVCacheBlock]`, which is always appended to in token order (new requests start empty, then blocks are added sequentially — `single_type_kv_cache_manager.py:212-251`). The trailing slots in the last block are handled by truncating to `num_tokens` in `extract_from_kv_cache`.
> 
> Re: hybrid models — mamba layers don't use this KV cache at all (they have their own SSM state). The connector explicitly filters to only the `CacheOnlyAttentionLayer` group (`example_hidden_states_connector.py:244-250`), which is always a `FullAttentionSpec` layer with no sliding window, so `remove_skipped_blocks` is a no-op and there are no null blocks in the block list.
> 

We always use our dummy layer to store the hidden states which is always a "Full attention" layer. If mamba or other layers are used, those will have separate storage from the dummy cacheonlylayer, so there shouldn't be any concern there.

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py (hunk 4 above)
Comment (fynnsu) [reply vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:478]: More claude (tldr: there's a separate list for unconfirmed spec tokens):

> This is safe — `all_token_ids` never contains rejected draft tokens. The two lists are kept separate:
> 
> - **Draft tokens** go into `request.spec_token_ids` (`request.py:148`), which is a separate list that gets cleared each step (`scheduler.py:531`).
> - **Accepted tokens** go through `RejectionSampler.parse_output` (`gpu_model_runner.py:301`), which strips rejected tokens from `sampled_token_ids` before returning. Only the accepted tokens + bonus token make it back to the scheduler, where `_update_request_with_output` (`scheduler.py:1674`) appends them to `_all_token_ids` via `append_output_token_ids`.
> 
> So by the time `request_finished` reads `request.all_token_ids`, it only contains prompt + accepted output tokens — each of which had a forward pass and therefore has a cached hidden state. The `[:-1]` trim on line 467 removes the final token (last accepted output), which was the model's output but never an input to a forward pass.
>

File: vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py [hunk 5]
```diff
@@ -218,42 +207,26 @@ def start_load_kv(self, *args, **kwargs: Any) -> None:
...
+        metadata = self._get_connector_metadata()
+        if not isinstance(metadata, ExampleHiddenStatesConnectorMetadata):
+            return
+        for req_id, filename in metadata.new_req_filenames.items():
+            if req_id in self._lock_fds:
+                continue
+            lock_path = filename + ".lock"
+            os.makedirs(os.path.dirname(lock_path), exist_ok=True)
+            lock_fd = os.open(lock_path, os.O_CREAT | os.O_WRONLY | os.O_TRUNC, 0o644)
+            fcntl.flock(lock_fd, fcntl.LOCK_EX)
```
Comment (shanjiaz) [inline vllm/distributed/kv_transfer/kv_connector/v1/example_hidden_states_connector.py:234]: Can we make sure only rank 0 is creating the lock file? Running with tp=4 I got
```
(Worker_TP1 pid=1714582) Creating lock for req_id=cmpl-af224c9ea1b4c618-0-85b9157b at /tmp/hidden_states_shanjiaz/cmpl-af224c9ea1b4c618-0-85b9157b.safetensors.lock
(Worker_TP0 pid=1714581) Creating lock for req_id=cmpl-af224c9ea1b4c618-0-85b9157b at /tmp/hidden_states_shanjiaz/cmpl-af224c9ea1b4c618-0-85b9157b.safetensors.lock
(Worker_TP3 pid=1714584) Creating lock for req_id=cmpl-af224c9ea1b4c618-0-85b9157b at /tmp/hidden_states_shanjiaz/cmpl-af224c9ea1b4c618-0-85b9157b.safetensors.lock
(Worker_TP2 pid=1714583) Creating lock for req_id=cmpl-af224c9ea1b4c618-0-85b9157b at /tmp/hidden_states_shanjiaz/cmpl-af224c9ea1b4c618-0-85b9157b.safetensors.lock
```
and they get stuck in a deadlock.

**Review by shanjiaz — CHANGES_REQUESTED** [review]

**Comment by fynnsu** [comment]
@mgoin Updated a few things. Should be ready for re-review.

Changes:
- Added forced chunked prefill test
- Added test for tp2 (requested by @shanjiaz since this was broken). Note this test isn't actually running in CI but I confirmed it passes locally
- Fixed tp>1 case (only tp rank0 saves hidden states, other ranks just pass)
- Fixed some issues with multiprocessing when running multiple tests back to back
- Updated tests to use `example_hidden_states_connector.load_hidden_states` which waits for lock before loading data (requirement for async)

On that last note, not sure why this wasn't flagged when the original async pr was merged, ~~I'm guessing these tests aren't being triggered in CI. I will look into this~~ Looks like they're running. Likely just a race condition with whether the file gets generated in time for us to try to load it.

**Comment by fynnsu** [comment]
~~Okay, noted the failure on [buildkite/ci/pr/extract-hidden-states-integration](https://buildkite.com/vllm/ci/builds/70637#019ea942-1e21-4de0-8cd6-010ee1c83fec) is related. Looks like an issue with the multi-processing setup, will fix.~~

Should be resolved. Will check to see if CI passes

**Review by mgoin — APPROVED** [review]

**Comment by fynnsu** [comment]
@mgoin Thanks for the review. ~~The one failing test is unrelated.~~

~~This pr is currently based off commit 99ef652907932fe6ba1fb28fa8bfbafeea8ed317 on the main branch which was merged on June 4, 2026. In the screenshot below, you can see the test was failing at that time.~~

Screenshot 2026-06-10 at 9 39 21 AM

~~I think this should be good to merge, but I can also rebase/merge main and re-run CI if you prefer.~~


All tests are passing after merges.

**Review by shanjiaz — APPROVED** [review]

**Review by robertgshaw2-redhat — APPROVED** [review]
